plugins/soundsourcem4a/audiosourcem4a.cpp

```diff
@@ -0,0 +1,329 @@
+        pDecoderConfig->outputFormat = FAAD_FMT_FLOAT; /* 32-bit float */
+        pDecoderConfig->downMatrix = 1; /* 5.1 -> stereo */
+        pDecoderConfig->defObjectType = LC;
+        if (!NeAACDecSetConfiguration(m_hDecoder, pDecoderConfig)) {
+            qWarning() << "Failed to configure AAC decoder!";
+            return ERR;
+        }
+    }
+
+    u_int8_t* configBuffer = NULL;
+    u_int32_t configBufferSize = 0;
+    if (!MP4GetTrackESConfiguration(m_hFile, m_trackId, &configBuffer,
+            &configBufferSize)) {
+        /* failed to get mpeg-4 audio config... this is ok.
+         * NeAACDecInit2() will simply use default values instead.
+         */
+        qWarning()
+                << "Failed to read the MP4 audio configuration. Continuing with default values.";
+    }
+
+    SAMPLERATE_TYPE sampleRate;
+    unsigned char channelCount;
+    if (0 > NeAACDecInit2(m_hDecoder, configBuffer, configBufferSize,
+                    &sampleRate, &channelCount)) {
+        free(configBuffer);
+        qWarning() << "Failed to initialize the AAC decoder!";
+        return ERR;
+    } else {
+        free(configBuffer);
+    }
+
+    setChannelCount(channelCount);
+    setFrameRate(sampleRate);
+    setFrameCount(m_maxSampleBlockId * kFramesPerSampleBlock);
+
+    // Allocate one block more than the number of sample blocks
+    // that are prefetched
+    const SampleBuffer::size_type prefetchSampleBufferSize =
+            (kNumberOfPrefetchSampleBlocks + 1) * frames2samples(kFramesPerSampleBlock);
+    m_prefetchSampleBuffer.resize(prefetchSampleBufferSize);
+
+    m_curSampleBlockId = MP4_INVALID_SAMPLE_ID;
+    m_curFrameIndex = getFrameCount();
+    // Seek to the beginning of the file
+    if (0 != seekSampleFrame(0)) {
+        qWarning() << "Failed to seek to the beginning of the file!";
+        return ERR;
+    }
+
+    return OK;
+}
+
+void AudioSourceM4A::close() {
+    if (m_hDecoder) {
+        NeAACDecClose(m_hDecoder);
+        m_hDecoder = NULL;
+    }
+    if (MP4_INVALID_FILE_HANDLE != m_hFile) {
+        MP4Close(m_hFile);
+        m_hFile = MP4_INVALID_FILE_HANDLE;
+    }
+    m_trackId = MP4_INVALID_TRACK_ID;
+    m_maxSampleBlockId = MP4_INVALID_SAMPLE_ID;
+    m_curSampleBlockId = MP4_INVALID_SAMPLE_ID;
+    m_inputBuffer.clear();
+    m_inputBufferOffset = 0;
+    m_inputBufferLength = 0;
+    m_curFrameIndex = 0;
+    reset();
+}
+
+bool AudioSourceM4A::isValidSampleBlockId(MP4SampleId sampleBlockId) const {
+    return (sampleBlockId >= kMinSampleBlockId) && (sampleBlockId <= m_maxSampleBlockId);
+}
+
+AudioSource::diff_type AudioSourceM4A::seekSampleFrame(diff_type frameIndex) {
+    if (m_curFrameIndex != frameIndex) {
+        const MP4SampleId sampleBlockId = kMinSampleBlockId
+                + (frameIndex / kFramesPerSampleBlock);
+        if (!isValidSampleBlockId(sampleBlockId)) {
+            return m_curFrameIndex;
+        }
+        if ((frameIndex < m_curFrameIndex) ||
```

Comment:
The first condition seams to be redundant. 
If I get this correct, every backward seek Reset the decoder. 
I think have to check if we leave the current SampleBlock.